src/errors/tags.ts

```diff
@@ -0,0 +1,18 @@
+import {EdgeDBError, Tag} from "./base";
+import {ErrorType} from "./base";
+import * as errors from "./index";
+
+class ListTag implements Tag {
+  classes: ErrorType[];
+  constructor(classes: ErrorType[]) {
+    this.classes = classes;
+  }
+  isMatching(err: EdgeDBError): boolean {
+    return this.classes.some((type) => err instanceof type);
+  }
+}
+
+export const SHOULD_RECONNECT = new ListTag([
+  errors.ConnectionFailedTemporarilyError,
+  errors.ConnectionTimeoutError,
+]);
```

Comment:
IMO we need to implement this differently.

I'd modify the current `edb/api/errors.txt` to include optional string tags, like

```
0x_05_02_00_00   IntegrityError                   repeatable, other_tag
0x_05_02_00_01   ConstraintViolationError
0x_05_02_00_02   CardinalityViolationError
0x_05_02_00_03   MissingRequiredError

0x_05_03_00_00   TransactionError                 repeatable
0x_05_03_00_01   TransactionSerializationError
0x_05_03_00_02   TransactionDeadlockError
```

and modify our JS errors code gen to include them in error classes, and also create a union string type of all available tags.

And then the client API will be

```
catch (e) {
   if (e instanceof EdgeDBError && e.hasTag('repeatable')) { ... }
}
```

This would align better with https://github.com/edgedb/edgedb/discussions/1849 and the idea that "tags" will be implemented via type annotations.

Reply:
I think tags have to be constants somewhere, not arbitrary strings. At least for easy discovery and to prevent typos. And I think we should postpone custom tags to later RFC when we implement custom errors.

Codegen for them is fine. But I'd rather make them opaque things for now, instead of hardcoding specific strings. I'm not sure I've achieved the goal in current implementation, though, as privacy rules of TypeScript are a bit unclear for me now. Also we can probably make them Symbols in JS?

Reply:
> I think tags have to be constants somewhere, not arbitrary strings. At least for easy discovery and to prevent typos. 

That's why I mentioned union string type. Basically you'll reflect all tags into

```typescript
type ErrorTag = 'repeatable' | 'retriable';

...

class Error {
  hasTag(tag: ErrorTag): bool;
}
```

We can do the same in Python.

> And I think we should postpone custom tags to later RFC when we implement custom errors.

Yes, but I'd like the API to be forward compatible. Annotations are the most straightforward and already supported way of implementing tags.

We can also reflect tags to sentinels:

```
const Repeatable = Symbol();
const Retriable = Symbol();
type ErrorTag = Repeatable | Retriable;

...

class Error {
  hasTag(tag: ErrorTag): bool;
}
```

I'd be OK with that; the benefit is that this would be safer to use without typing. In this case you wouldn't need to use `instanceof` and `map` -- just a `Set().has(tag)` would work.

Reply:
Let's do the sentinel approach (`const Repeatable = Symbol();`)

Reply:
Fixed.